src/tests/test_prep.py

```diff
@@ -6,6 +6,129 @@
 import hivpy.column_names as col
 from hivpy.common import SexType, date, timedelta
 from hivpy.population import Population
+from hivpy.prep import PrEPType
+
+
+def reset_prep_willingness_cols(pop: Population):
+    pop.set_present_variable(col.PREP_ORAL_PREF, 0)
+    pop.set_present_variable(col.PREP_CAB_PREF, 0)
+    pop.set_present_variable(col.PREP_LEN_PREF, 0)
+    pop.set_present_variable(col.PREP_VR_PREF, 0)
+    pop.set_present_variable(col.PREP_PREF_RANKED, None)
+    pop.set_present_variable(col.PREP_ORAL_WILLING, False)
+    pop.set_present_variable(col.PREP_CAB_WILLING, False)
+    pop.set_present_variable(col.PREP_LEN_WILLING, False)
+    pop.set_present_variable(col.PREP_VR_WILLING, False)
+    pop.set_present_variable(col.PREP_ANY_WILLING, False)
+
+
+def test_prep_willingness():
+    N = 100
+    pop = Population(size=N, start_date=date(1999, 1, 1))
+    pop.data[col.AGE] = 20
+    pop.data[col.VIRAL_LOAD] = 10000
+    # make all prep types available
+    pop.prep.date_prep_intro = [date(2000), date(3000), date(4000), date(5000)]
+    # adjust chances of higher preference
+    pop.prep.prep_oral_pref_beta = 3
+    pop.prep.prep_cab_pref_beta = 3.3
+    pop.prep.prep_len_pref_beta = 3.3
+    pop.prep.prep_vr_pref_beta = 2.9
+    # vl prevalence accounted for
+    pop.prep.vl_prevalence_affects_prep = True
+    pop.prep.vl_prevalence_prep_threshold = 0.5
+
+    # no willingness before prep intro date
+    pop.prep.prep_willingness(pop)
+    assert sum(pop.data[col.PREP_ANY_WILLING]) == 0
+
+    # willingness calculated for over 15s
+    pop.date = date(2000, 1, 1)
+    pop.prep.prep_willingness(pop)
+    # some oral willingness established
+    assert sum(pop.data[col.PREP_ORAL_WILLING]) > 0
+    assert sum(pop.data[col.PREP_CAB_WILLING]) == 0
+    assert sum(pop.data[col.PREP_LEN_WILLING]) == 0
+    assert sum(pop.data[col.PREP_VR_WILLING]) == 0
+    # check for highest preference
+    assert all([i[0] is PrEPType.Oral for i in pop.data[col.PREP_PREF_RANKED]])
```

Comment:
```suggestion
    # willingness calculated for over 15s
    pop.date = date(2000, 1, 1)
    pop.prep.prep_willingness(pop)
    
    under_15s = pop.get_sub_pop(COND(col.AGE, op.lt, 15))
    over_15s = pop.get_sub_pop(COND(col.AGE, op.ge, 15))
    # some oral willingness established
    assert sum(pop.get_variable(col.PREP_ORAL_WILLING, under_15s)) == 0
    assert sum(pop.get_variable(col.PREP_CAB_WILLING, under_15s)) == 0
    assert sum(pop.get_variable(col.PREP_LEN_WILLING, under_15s)) == 0
    assert sum(pop.get_variable(col.PREP_VR_WILLING, under_15s)) == 0
    
    assert sum(pop.get_variable(col.PREP_ORAL_WILLING, over_15s)) > 0
    assert sum(pop.get_variable(col.PREP_CAB_WILLING, over_15s)) == 0
    assert sum(pop.get_variable(col.PREP_LEN_WILLING, over_15s)) == 0
    assert sum(pop.get_variable(col.PREP_VR_WILLING, over_15s)) == 0
    # check for highest preference
    assert all([i[0] is PrEPType.Oral for i in pop.data[col.PREP_PREF_RANKED]])
```

Checking for over and under 15 behaviour